Rails Course: Update for Turbo Drive forms compatibility (#23797)
TheOdinProject/curriculum: rails_programming/assets_and_navigation/turbo_drive.md

Turbo itself is also an umbrella term for several different techniques for creating fast and modern web applications. The technique we're going to focus on here is the one whose behaviour is part of Rails' core behaviour, Turbo Drive. The other techniques we'll cover in the Advanced Turbo lesson later.

Links and Forms are the bedrock of most web applications. You need links to navigate around your site and forms to allow users to submit information to you. Turbo Drive is how Rails handles these in your applications.

If a lot of what follows seems confusing, don't worry! A lot of Turbo Drive happens under the hood and therefore it means we often need to discuss the concept rather than the implementation. This can make it harder initially to grok, but luckily it does also mean you can often just go about your business and never have to actively think about it.

### Learning Outcomes

*Look through these now and then use them to test yourself after doing the assignment*

* What is Turbolinks?
* What is the purpose of Turbo Drive?
* How does Turbo Drive handle page navigation?
* What are the two kinds of page visit types in Turbo?
* How can you disable Turbo Drive?

### Turbolinks

> Before we can talk about Turbo, we should briefly mention its predecessor, Turbolinks. Turbolinks is **no longer in active development**! Despite that, it's still worth mentioning as it has been around for a while (its first major release was 2013) and it's possible that you will encounter existing applications or online articles that mention Turbolinks. The goal of Turbolinks was to capture the request for any clicked links in your web application and, if that link was to another page in your web app and not to an outside resource, then Turbolinks would keep the current page instance alive and only swap out the content between the `<body>` tags of the document. Because it didn't do a full page redirect it was faster because it didn't have to attempt to download any resources in the head part of the document when they were unlikely to have unchanged.

Reviewer: I think it may be a little confusing to start with turbolinks. Imo this would fit better at the end as an aside. 

I think we could reshuffle things a little here by starting with your paragraph below on what turbo drive is in and then going into the good stuff you've written here about how it intercepts requests to only reload the parts it needs to. That answers the "What is the thing and why does it exist?" question that gets readers into the write frame.

  <%= link_to "baz", "qux", data: { turbo: "true" } %>
</div>
~~~

In the above example, we created a parent div with `data-turbo="false"`, which disables Turbo Drive on all elements inside of it. However, we also added `data: { turbo: "true" }` to the second `link_to`. This will turn Turbo Drive back on for that particular element.

This showcases that you can disable and re-enable Turbo Drive in regions of your website and select specific elements to enable or disable it on. Remember that when using ERB tags (like `link_to`) we use the syntax style of `data: { foo: "bar" }` rather than the HTML style of `data-foo="bar"`

### Forms

We cover Rails forms much deeper later, but because they are entwined with Turbo Drive we do need to cover some common information that will come together in the forms lessons that follow.

If you don't understand some of the information as it is explained below just wait until the form basics lesson and it should all come together.

#### Form Submissions

While most link requests are issued as GET requests, forms are different. You often submit forms as a POST, PATCH or DELETE request in order to manipulate some data that lives on the server.

> Turbo Drive intercepts all Form Submissions by default and submits them itself to the server on your behalf. What you must keep in mind going forwards into the form lessons is that when Turbo expects the server usually to return an HTTP status of HTTP 303. This is a redirect response. Think of it like this, after you submit a form usually you want to do be directed somewhere afterwards. That could be back to the same page with a blank form ready for another submission, or it could be to view the resource you just created with the form, but you expect to be redirected away from the current page with the completed form on it. There are two exceptions to this expectation

Reviewer: ```suggestion
Turbo Drive intercepts all Form Submissions by default and submits them to the server on your behalf. As we advance into the form lessons, you must remember that Turbo expects the server to return an HTTP status of HTTP 303 or, in other words, a redirect.

Think of it like this: after you submit a form, you usually want to be directed somewhere afterwards. That could be back to the same page with a blank form ready for another submission, or it could be to view the resource you just created with the form, but you expect to be redirected away from the current page with the completed form on it. There are two exceptions to this expectation.
```
